include/ensmallen_bits/pso/pso.hpp

```diff
@@ -0,0 +1,138 @@
+/**
+ * @file pso.hpp
+ * @author Chintan Soni
+ * @author Suryoday Basak
+ *
+ * Particle swarm optimization.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_PSO_PSO_HPP
+#define ENSMALLEN_PSO_PSO_HPP
+
+#include "update_policies/lbest_update.hpp"
+#include "init_policies/default_init.hpp"
+#include "init_policies/constrained_init.hpp"
+#include <iostream>
+
+namespace ens {
+
+/**
+ * EXTREMELY DETAILED DESCRIPTION OF THE WORKING OF PSO.
+ *
+ * For Particle Swarm Optimization to work, a FunctionType template parameter is
+ * required.
+ * This class must implement the following function:
+ *
+ *   double Evaluate(const arma::mat& coordinates);
+ *   void Gradient(const arma::mat& coordinates,
+ *                 arma::mat& gradient);
+ */
+template<typename VelocityUpdatePolicy = LBestUpdate,
+         typename InitPolicy = DefaultInit>
+class PSOType
+{
+ public:
+  /**
+   * Construct the particle swarm optimizer with the given function and
+   * parameters. The defaults here are not necessarily good for the given
+   * problem, so it is suggested that the values used be tailored to the task
+   * at hand.
+   *
+   * @param numParticles Number of particles in the swarm.
+   * @param maxIterations Number of iterations allowed.
+   * @param exploitationFactor Influence of the personal best of the particle.
+   * @param explorationFactor Influence of the neighbours of the particle.
+   * @param enableGradientDescent Enable the use of gradient descent optimizer.
```

Comment:
Looks like the next three aren't used, and there is no description for ``velocityUpdatePolicy`` and ``initPolicy``.

Reply:
Yes, I'll remove these and as a part of the 'detailed description', I will add descriptions to each of the literals.

Reply:
I've done this.